### DISCUSSION TITLE
Close error on httpServer close when attaching multiple Server instance to same httpServer

### DISCUSSION THREAD

**Discussion by UnrealDoggie** [socketio/socket.io#4002] [Q&A] [answered]
Hi,

I created multiple Server instances specifying different paths (not namespaces)  like
/Company1 , /Company2  & attaching them to the same httpServer running
the same port. 
Upon server shutdown, I wish to close each of the instances calling ioServer.close().
However, because it also calls the underlying httpServer.close() , I run into httpServer.close error
ERR_SERVER_NOT_RUNNING] later.
My question is 
Is it sufficient to call 
ioServer?.disconnectSockets(true);
ioServer?.engine.close();

to close socket.io & close httpServer explicitly?

async shutdown(): Promise<void> {
console.log("Calling shutdown");
    for (const key in this.companies) {
      const ioServer = this.companies[key].getSocketServer();
       // close calls the httpServer to close as well!
       ioServer?.close();
    }
    // destroy all active connections
    for (const connection of this.httpConnections) {
      connection.destroy();
      this.httpConnections.delete(connection);
    }
    await new Promise((resolve, reject) => {
      this.httpServer.close((err) => {
        if (err) {
          reject(err);
        } else {
          console.log("http connections all closed");
          resolve(true);
        }
      });
    });
}

Thanks in advance!

**Comment by darrachequesne** [comment, marked as answer]
Hi!

> Is it sufficient to call
> ioServer?.disconnectSockets(true);
> ioServer?.engine.close();
> to close socket.io & close httpServer explicitly?

Yes. I would add a `local` flag (`ioServer?.local.disconnectSockets(true)`), if you are in a multi-server setup.

That being said, I'm wondering whether we should actually close the underlying HTTP server if it was provided by the user. WDYT?

**Reply by UnrealDoggie** [reply]
Hi Damien,

Is having multiple paths /Company1 , /Company2 on the same http server recommended in the first place? I am not actually running multi nodes with sticky session.

I am not getting a clean teardown in jest (even ioServer?.local.disconnectSockets(true))
"Jest has detected the following 1 open handle potentially keeping Jest from exiting: this.server.listen(this.port, () => {"
It does not happen if I create only one /Company1.

> That being said, I'm wondering whether we should actually close the underlying HTTP server if it was provided by the user. WDYT?

That might be worth a try or pass in a flag to so that httpServer.close is not called prematurely. Is there something important I skip by not calling ioServer?.close ?
Coz the ._onClose will not be called then.

Thanks,
Eugene

**Reply by nwalters512** [reply]
Hey @darrachequesne, I came across this reply while trying to figure out if it's possible to shut down socket.io without also closing the underlying HTTP server.

> That being said, I'm wondering whether we should actually close the underlying HTTP server if it was provided by the user. WDYT?

I could understand closing the server if one is made by socket.io on behalf of the caller, but IMO it's extremely surprising to me that socket.io would try to close a resource that it itself didn't create.

**Reply by darrachequesne** [reply]
@nwalters512 I agree. We'll change this behavior in the next major version.

**Reply by nwalters512** [reply]
@darrachequesne  that's great news! Is there an issue tracking that change that I can subscribe to? Otherwise I'd be happy to open one.

**Reply by darrachequesne** [reply]
@nwalters512 here we go: https://github.com/socketio/socket.io/discussions/4529
